### PR TITLE
feat(oauth2): added support

### PR DESCRIPTION
**Pull request by alesbrelih**
Hi!

Thanks for this generation, after quite a search it seems like this generator actually works! And it has nice golang syntax behind it.

Because the only limitation I have with this generation is the lack of oauth2 security provider I decided I wanted to contribute to this project.

I wanted to implement it as simple as possible.

Thanks!

### CODE REVIEW

**Comment by alesbrelih** [comment]
Forgot to run lint task, now it should be fixed.

File: gen/_template/security.tmpl [hunk 1]
```diff
@@ -29,6 +29,18 @@ func findAuthorization(h http.Header, prefix string) (string, bool) {
...
+{{- if $s.Format.IsOAuth2Security }}
+var oauth2Scopes = map[string][]string {
+{{- range $operationName, $scopes := $s.Scopes }}
+  {{ quote $operationName }}: []string{
+    {{- range $scope := $scopes }}
+      {{ quote $scope }},
+    {{- end}}
+  },
+{{- end }}
+}
```
Comment (shadowspore) [inline gen/_template/security.tmpl:41]: According to our [.editorconfig](https://github.com/ogen-go/ogen/blob/main/.editorconfig#L11-L13) we use tabs in template files.

File: gen/_template/security.tmpl [hunk 2]
```diff
@@ -74,6 +86,13 @@ func (s *Server) security{{ $s.Type.Name }}(ctx context.Context, operationName s
 			return ctx, false, nil
 		}
 		t.Token = token
+	{{- else if $s.Format.IsOAuth2Security }}
+		token, ok := findAuthorization(req.Header, "Bearer")
+		if !ok {
+			return ctx, false, nil
+		}
+		t.Token = token
+		t.Scopes = oauth2Scopes[operationName]
```
Comment (shadowspore) [inline gen/_template/security.tmpl:95]: @tdakkota is an arbitrary `operationName` (not present in the spec and code) possible here?
If so, we need to handle this case by making sure that the oauth scopes for `operationName` exist in the map.

File: gen/_template/security.tmpl (hunk 1 above)
Comment (alesbrelih) [reply gen/_template/security.tmpl]: Changed :+1:

File: gen/_template/security.tmpl (hunk 2 above)
Comment (alesbrelih) [reply gen/_template/security.tmpl:95]: I was thinking of this case but then realized that with go and how it handles maps its not a problem if it won't exist. If map key doesn't exist it returns just a default version of the value type. So I thought it's not really needed to add extra logic.

Example:
https://go.dev/play/p/autilf-GwdK

Comment (alesbrelih) [reply gen/_template/security.tmpl:95]: Meaning if operationName doesn't exist it Scopes will be set to an empty list.

Comment (shadowspore) [reply gen/_template/security.tmpl:95]: Sure, but I'm more concerned about is it ok to pass empty oauth scopes to security handler?
Does oauth2 spec allow that?

Comment (alesbrelih) [reply gen/_template/security.tmpl:95]: Yeah was wondering the same thing when implementing. According to the spec here: https://swagger.io/docs/specification/authentication/oauth2/.

```
No Scopes
Scopes are optional, and your API may not use any. In this case, specify an empty object {} in the scopes definition, and an empty list of scopes [] ...
```

Meaning its up to the spec security implementator to decide if he wants to use scopes or not 🤔

Comment (shadowspore) [reply gen/_template/security.tmpl:95]: Seems fine - https://swagger.io/docs/specification/authentication/oauth2/

**Comment by shadowspore** [comment]
@tdakkota ptal

**Review by tdakkota — APPROVED** [review]

**Pull request merged**